File: ai-service/services/family_portfolio.py
from __future__ import annotations

from models import AssetBase, HoldingBase, InvestmentThesisBase, PerformanceHistoryBase, PortfolioBase
from repositories import FamilyPortfolioRepository


class FamilyPortfolioService:
    def __init__(self, repository: FamilyPortfolioRepository | None = None):
        self.repository = repository or FamilyPortfolioRepository()

    def list_portfolios(self) -> dict:
        portfolios = self.repository.list_portfolios()
        return {"items": [portfolio.model_dump(mode="json") for portfolio in portfolios]}

    def create_portfolio(self, payload: PortfolioBase) -> dict:
        portfolio = self.repository.create_portfolio(payload)
        return portfolio.model_dump(mode="json")

    def update_portfolio(self, portfolio_id: int, payload: PortfolioBase) -> dict | None:
        portfolio = self.repository.update_portfolio(portfolio_id, payload)
        return portfolio.model_dump(mode="json") if portfolio else None

    def delete_portfolio(self, portfolio_id: int) -> bool:
        return self.repository.delete_portfolio(portfolio_id)

    def list_assets(self) -> dict:
        assets = self.repository.list_assets()
        return {"items": [asset.model_dump(mode="json") for asset in assets]}

    def create_asset(self, payload: AssetBase) -> dict:
        asset = self.repository.create_asset(payload)
        return asset.model_dump(mode="json")

    def update_asset(self, asset_id: int, payload: AssetBase) -> dict | None:
        asset = self.repository.update_asset(asset_id, payload)
        return asset.model_dump(mode="json") if asset else None

    def delete_asset(self, asset_id: int) -> bool:
        return self.repository.delete_asset(asset_id)

    def create_holding(self, payload: HoldingBase) -> dict:
        holding = self.repository.create_holding(payload)
        return holding.model_dump(mode="json")

    def update_holding(self, holding_id: int, payload: HoldingBase) -> dict | None:
        holding = self.repository.update_holding(holding_id, payload)
        return holding.model_dump(mode="json") if holding else None

    def delete_holding(self, holding_id: int) -> bool:
        return self.repository.delete_holding(holding_id)

    def create_investment_thesis(self, payload: InvestmentThesisBase) -> dict:
        thesis = self.repository.create_investment_thesis(payload)
        return thesis.model_dump(mode="json")

    def update_investment_thesis(self, thesis_id: int, payload: InvestmentThesisBase) -> dict | None:
        thesis = self.repository.update_investment_thesis(thesis_id, payload)
        return thesis.model_dump(mode="json") if thesis else None

    def delete_investment_thesis(self, thesis_id: int) -> bool:
        return self.repository.delete_investment_thesis(thesis_id)

    def create_performance_history(self, payload: PerformanceHistoryBase) -> dict:
        history = self.repository.create_performance_history(payload)
        return history.model_dump(mode="json")

    def update_performance_history(
        self, history_id: int, payload: PerformanceHistoryBase
    ) -> dict | None:
        history = self.repository.update_performance_history(history_id, payload)
        return history.model_dump(mode="json") if history else None

    def delete_performance_history(self, history_id: int) -> bool:
        return self.repository.delete_performance_history(history_id)

    def get_portfolio_detail(self, portfolio_id: int) -> dict | None:
        aggregate = self.repository.get_portfolio_aggregate(portfolio_id)
        if aggregate is None:
            return None
        return aggregate.model_dump(mode="json")

File: frontend/src/api/familyPortfolio.js
import axios from 'axios'
import { useAuthStore } from '@/stores/auth'

const familyPortfolioApi = axios.create({
  baseURL: '/ai-api/family-portfolio',
  timeout: 20000,
  headers: {
    'Content-Type': 'application/json'
  }
})

function getAuthHeaders() {
  const authStore = useAuthStore()
  return authStore.apiKey
    ? {
        'X-API-Key': authStore.apiKey
      }
    : {}
}

export const portfolioApi = {
  listPortfolios() {
    return familyPortfolioApi.get('/portfolios')
  },
  getPortfolioDetail(portfolioId) {
    return familyPortfolioApi.get(`/portfolios/${portfolioId}`)
  },
  listAssets() {
    return familyPortfolioApi.get('/assets')
  }
}

export const adminPortfolioApi = {
  listPortfolios() {
    return familyPortfolioApi.get('/portfolios')
  },
  getPortfolioDetail(id) {
    return familyPortfolioApi.get(`/portfolios/${id}`)
  },
  createPortfolio(data) {
    return familyPortfolioApi.post('/portfolios', data, { headers: getAuthHeaders() })
  },
  updatePortfolio(id, data) {
    return familyPortfolioApi.put(`/portfolios/${id}`, data, { headers: getAuthHeaders() })
  },
  deletePortfolio(id) {
    return familyPortfolioApi.delete(`/portfolios/${id}`, { headers: getAuthHeaders() })
  },
  listAssets() {
    return familyPortfolioApi.get('/assets')
  },
  createAsset(data) {
    return familyPortfolioApi.post('/assets', data, { headers: getAuthHeaders() })
  },
  updateAsset(id, data) {
    return familyPortfolioApi.put(`/assets/${id}`, data, { headers: getAuthHeaders() })
  },
  deleteAsset(id) {
    return familyPortfolioApi.delete(`/assets/${id}`, { headers: getAuthHeaders() })
  },
  createHolding(data) {
    return familyPortfolioApi.post('/holdings', data, { headers: getAuthHeaders() })
  },
  updateHolding(id, data) {
    return familyPortfolioApi.put(`/holdings/${id}`, data, { headers: getAuthHeaders() })
  },
  deleteHolding(id) {
    return familyPortfolioApi.delete(`/holdings/${id}`, { headers: getAuthHeaders() })
  },
  createInvestmentThesis(data) {
    return familyPortfolioApi.post('/investment-theses', data, { headers: getAuthHeaders() })
  },
  updateInvestmentThesis(id, data) {
    return familyPortfolioApi.put(`/investment-theses/${id}`, data, { headers: getAuthHeaders() })
  },
  deleteInvestmentThesis(id) {
    return familyPortfolioApi.delete(`/investment-theses/${id}`, { headers: getAuthHeaders() })
  },
  createPerformanceHistory(data) {
    return familyPortfolioApi.post('/performance-history', data, { headers: getAuthHeaders() })
  },
  updatePerformanceHistory(id, data) {
    return familyPortfolioApi.put(`/performance-history/${id}`, data, { headers: getAuthHeaders() })
  },
  deletePerformanceHistory(id) {
    return familyPortfolioApi.delete(`/performance-history/${id}`, { headers: getAuthHeaders() })
  }
}
